rig-sqlite/src/lib.rs

```diff
@@ -0,0 +1,360 @@
+use rig::embeddings::{DocumentEmbeddings, Embedding, EmbeddingModel};
+use rig::vector_store::{VectorStore, VectorStoreError, VectorStoreIndex};
+use rusqlite::OptionalExtension;
+use serde::Deserialize;
+use tokio_rusqlite::Connection;
+use tracing::{debug, info};
+
+#[derive(Debug)]
+pub enum SqliteError {
+    DatabaseError(Box<dyn std::error::Error + Send + Sync>),
+    SerializationError(Box<dyn std::error::Error + Send + Sync>),
+}
+
+#[derive(Clone)]
+pub struct SqliteVectorStore {
+    conn: Connection,
+}
+
+impl SqliteVectorStore {
+    pub async fn new(conn: Connection) -> Result<Self, VectorStoreError> {
+        debug!("Running initial migrations");
+        // Run migrations or create tables if they don't exist
+        conn.call(|conn| {
+            conn.execute_batch(
+                "BEGIN;
+                -- Document tables
+                CREATE TABLE IF NOT EXISTS documents (
+                    id INTEGER PRIMARY KEY AUTOINCREMENT,
+                    doc_id TEXT UNIQUE NOT NULL,
+                    document TEXT NOT NULL
+                );
+                CREATE INDEX IF NOT EXISTS idx_doc_id ON documents(doc_id);
+                CREATE VIRTUAL TABLE IF NOT EXISTS embeddings USING vec0(embedding float[1536]);
+
+                COMMIT;",
```

Comment:
This query hardcodes a couple of names and values such as `idx_doc_id`, `doc_id`, `1536`, etc. These should be configurable via the `new` (and potentially could be done via a builder for the vector store if you wanted to provide defaults). If you feel some should be hardcoded, then they should be constants but do provide a reason.

For the length of the embeddings, `1536`, we would use `EmbeddingModel::ndims` to configure this. This would mean `new` would also require an `EmbeddingModel` as well (and would need to be generic over one).

Note: in the embeddings revamp, the actual `VectorStore` general trait is gone so that would all collapse into your `SqliteVectorStore`.

Reply:
I'm not sure extracting these to be configurable is worth the additional complexity, since the interface encapsulates those details, and imo reduces the readability of the queries when there is a lot of string interpolation going on.

Either way, I think it might be better to save it for the embeddings refactor, since that will impact this pretty significantly from what I can tell.